node/alphanet/src/block_signer.rs

```diff
@@ -0,0 +1,251 @@
+use primitives::chain::SignedShardBlock;
+use primitives::beacon::SignedBeaconBlock;
+use tokio::timer::Delay;
+use primitives::types::AuthorityId;
+use primitives::types::PartialSignature;
+use primitives::types::BlockIndex;
+use futures::sync::mpsc::{Receiver, Sender};
+use futures::Future;
+use std::time::Instant;
+use std::time::Duration;
+use futures::stream::Stream;
+use futures::Async;
+use futures::Poll;
+use nightshade::nightshade_task::GossipBody;
+use nightshade::nightshade_task::Gossip;
+use futures::sink::Sink;
+use primitives::block_traits::SignedBlock;
+
+const COOLDOWN_MS: u64 = 200;
+
+struct Confirmation(SignedBeaconBlock, SignedShardBlock);
+
+enum BlockSignerGossipBody {
+    Request,
+    Reply(PartialSignature),
+}
+
+enum BlockSignerControl {
+    Reset(SignedBeaconBlock, SignedShardBlock, BlockIndex),
+    Stop,
+}
+
+/// TODO: Plug this task in alphanet.
+
+/// TODO: Gossips are not signed, because we are only exchanging BLS signature. One edge of attack
+/// might be spam a node with invalid BLS signature (BLS is slow and checking for invalid signature
+/// is expensive).
+struct BlockSignerGossip {
+    receiver_id: AuthorityId,
+    sender_id: AuthorityId,
+    body: BlockSignerGossipBody,
+    block_index: BlockIndex,
+}
+
+struct BlockSignerTask {
+    owner_id: AuthorityId,
+
+    beacon_block: Option<SignedBeaconBlock>,
+    shard_block: Option<SignedShardBlock>,
+
+    num_authorities: Option<usize>,
+    block_index: Option<BlockIndex>,
+
+    control_rx: Receiver<BlockSignerControl>,
+    inc_gossips_rx: Receiver<BlockSignerGossip>,
+    out_gossips_tx: Sender<BlockSignerGossip>,
+    confirmation_tx: Sender<Confirmation>,
+
+    cooldown_delay: Option<Delay>,
+    has_confirmed: bool,
+}
+
+impl BlockSignerTask {
```

Comment:
This seems like a lot of code, I was imagining this to be just an extension of block producer module:
```
fn spawn_block_producer( ...) {
   consensus.for_each(|..| {
        if let (new_blocks) = try_produce_block(...) {
             client.unsigned_blocks.insert(hash or index, new_blocks)
             my_sig = new_blocks.sign(client.signer)
             send to random authorities in out_sig_gossips <- my_sig
        }
   })
   inc_sig_gossips.for_each(|..| {
       block = client.unsigied_blocks.entry(gossip.hash or index)
       block.add_signature(received sig that may contain a bunch of others)
       send to random authorities in out_sig_gossips <- block.sigs (total accumulated so far)
       if block.signature.num() > 2/3 num_authorities {
            client.unsigned_blocks.pop(index or hash) -> import to client properly
            send Reset to mempool_control
       }
   })
} 
```

so keep track in client of unsigned block(s) and the block's signature already maintains who's signatures we have.

Reply:
I agree with @ilblackdragon . This task should not exist on its own, because having more tasks with controls that are chained according to some logic is very error-prone. At the minimum it should be a part of producer.

However, I think we need to merge block importer, block production, mempool coros and this task into one, called client coro. This will also remove the mempool control and other synchronization stuff. Also, currently importer, production and mempool are semi-synchronized with each other anyway, I think they should be fully sequentialized through client coro, which will make code more diagnosable.

Reply:
`send to random authorities in out_sig_gossips <- my_sig`
@ilblackdragon all the point to do this is because if sending BLS signature fails, we need to request it again... and hence, have this "yet another gossip" layer to request, replay signatures! We just can't be optimistic about it and think that at least 2/3 + 1 will arrive. (We can use your approach for now).